src/drivers/amd/rembrandt/acp_sw_audio_dma.c

```diff
@@ -0,0 +1,640 @@
+
+#define SW0_BT_FIFO_SIZE		128
+#define SW0_BT_TX_FIFO_ADDR		(SW0_AUDIO_RX_FIFO_ADDR + SW0_AUDIO_FIFO_SIZE)
+#define SW0_BT_RX_FIFO_ADDR		(SW0_BT_TX_FIFO_ADDR + SW0_BT_FIFO_SIZE)
+
+#define SW0_HS_FIFO_SIZE		128
+#define SW0_HS_TX_FIFO_ADDR		(SW0_BT_RX_FIFO_ADDR + SW0_BT_FIFO_SIZE)
+#define SW0_HS_RX_FIFO_ADDR		(SW0_HS_TX_FIFO_ADDR + SW0_HS_FIFO_SIZE)
+
+#define SW1_FIFO_SIZE			128
+#define SW1_TX_FIFO_ADDR		(SW0_HS_RX_FIFO_ADDR + SW1_FIFO_SIZE)
+#define SW1_RX_FIFO_ADDR		(SW1_TX_FIFO_ADDR + SW1_FIFO_SIZE)
+
+static uint32_t sw_audio_buff_size_playback;
+static uint32_t sw_audio_buff_size_capture;
+
+struct sw_dev_register {
+	uint32_t	sw_dev_en;
+	uint32_t	sw_dev_en_status;
+	uint32_t	sw_dev_fifo_addr;
+	uint32_t	fifo_addr;
+	uint32_t	sw_dev_fifo_size;
+	uint32_t	fifo_size;
+	uint32_t	sw_dev_ringbuff_addr;
+	uint32_t	sw_dev_ringbuff_size;
+	uint32_t	sw_dev_dma_size;
+	uint32_t	sw_dev_dma_watermark;
+	uint32_t	sw_dev_dma_intr_status;
+	uint32_t	sw_dev_dma_intr_cntl;
+	uint32_t	statusindex;
+};
+
+struct sw_dev_register sw_dev[8];
+
+/* allocate next free DMA channel */
+static struct dma_chan_data *acp_dai_sw_audio_dma_channel_get(struct dma *dma,
+							      unsigned int req_chan)
+{
+	k_spinlock_key_t key;
+	struct dma_chan_data *channel;
+
+	key = k_spin_lock(&dma->lock);
+	if (req_chan >= dma->plat_data.channels) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_sw_audio_tr, "Channel %d not in range", req_chan);
+		return NULL;
+	}
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_sw_audio_tr, "channel already in use %d", req_chan);
+		return NULL;
+	}
+	atomic_add(&dma->num_channels_busy, 1);
+	channel->status = COMP_STATE_READY;
+	k_spin_unlock(&dma->lock, key);
+	return channel;
+}
+
+/* channel must not be running when this is called */
+static void acp_dai_sw_audio_dma_channel_put(struct dma_chan_data *channel)
+{
+	k_spinlock_key_t key;
+
+	notifier_unregister_all(NULL, channel);
+	key = k_spin_lock(&channel->dma->lock);
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+	k_spin_unlock(&channel->dma->lock, key);
+}
+
+static int acp_dai_sw_audio_dma_start(struct dma_chan_data *channel)
+{
+	uint32_t		sw0_audio_tx_en = 0;
+	uint32_t		sw0_audio_rx_en = 0;
+	uint32_t		acp_pdm_en;
+	int    tx_en_status;
+	int    rx_en_status;
+	uint32_t delay_cnt = 10000;
+	int i = 0;
+
+	sw0_audio_tx_en = 0;
+	sw0_audio_rx_en = 0;
+	for (i = 0; i < 8; i = i + 2) {
+		sw0_audio_tx_en |= io_reg_read((PU_REGISTER_BASE + sw_dev[i].sw_dev_en));
+		sw0_audio_rx_en |= io_reg_read((PU_REGISTER_BASE + sw_dev[i + 1].sw_dev_en));
+	}
+
+	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
+	if (!sw0_audio_tx_en && !sw0_audio_rx_en && !acp_pdm_en) {
+		/* Request SMU to set aclk to 600 Mhz */
+		acp_change_clock_notify(600000000);
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), ACP_ACLK_CLK_SEL);
+	}
```

Comment:
it's a bit odd to combine information from two separate dais here. This feels like there's a missing clock driver or something is provided by multiple requests. making clock decisions in a DMA is weird.

Reply:
Currently proper clock driver is not available for AMD platform, to support S0i3 and multiple pipelines it is required to set the clocks accordingly as per ACP architecture. Hence clock changes are done in DMA.